pkg/agent/openflow/pipeline.go

```diff
@@ -825,32 +966,187 @@ func (c *client) l3ToExternalFlows(nodeIP net.IP, localSubnet net.IPNet, outputP
 	return flows
 }
 
+// serviceLearnFlow generates the flow with learn action which adds new flows in
+// sessionAffinityTable according to the Endpoint selection decision.
+func (c *client) serviceLearnFlow(svcIP net.IP, svcPort uint16, protocol binding.Protocol, affinityTimeout uint16) binding.Flow {
+	learnFlowBuilder := c.pipeline[serviceLBTable].BuildFlow(priorityLow).
+		MatchRegRange(int(serviceLearnReg), marksRegServiceNeedLearn, serviceLearnRegRange).
+		MatchDstIP(svcIP).
+		Cookie(c.cookieAllocator.Request(cookie.Service).Raw())
+	learnFlowBuilderLearnAction := learnFlowBuilder.
+		Action().Learn(sessionAffinityTable, priorityNormal, affinityTimeout, 0, c.cookieAllocator.Request(cookie.Service).Raw()).
+		DeleteLearned()
+	if protocol == binding.ProtocolTCP {
+		learnFlowBuilder = learnFlowBuilder.MatchTCPDstPort(svcPort)
+		learnFlowBuilderLearnAction = learnFlowBuilderLearnAction.MatchLearnedTCPDstPort()
+	} else {
+		learnFlowBuilder = learnFlowBuilder.MatchUDPDstPort(svcPort)
+		learnFlowBuilderLearnAction = learnFlowBuilderLearnAction.MatchLearnedUDPDstPort()
+	}
+	return learnFlowBuilderLearnAction.
+		MatchLearnedDstIP().
+		MatchLearnedSrcIP().
+		LoadRegToReg(int(serviceIPReg), int(serviceIPReg), serviceIPRegRange, serviceIPRegRange).
+		LoadRegToReg(int(servicePortReg), int(servicePortReg), servicePortRegRange, servicePortRegRange).
+		LoadReg(int(serviceLearnReg), marksRegServiceHitted, serviceLearnRegRange).
+		Done().
+		Action().LoadRegRange(int(serviceLearnReg), marksRegServiceHitted, serviceLearnRegRange).
+		Action().GotoTable(endpointDNATTable).
+		Done()
+}
+
+// serviceLBFlow generates the flow which uses the specific group to do Endpoint
+// selection.
+func (c *client) serviceLBFlow(groupID binding.GroupIDType, svcIP net.IP, svcPort uint16, protocol binding.Protocol) binding.Flow {
+	lbFlowBuilder := c.pipeline[serviceLBTable].BuildFlow(priorityNormal)
+	if protocol == binding.ProtocolTCP {
+		lbFlowBuilder = lbFlowBuilder.MatchTCPDstPort(svcPort)
+	} else {
```

Comment:
Kubernetes also supports SCTP. So it is not accurate to use if TCP else UDP.

Reply:
Thanks, added.